### ISSUE TITLE
[🐛] Missing firebase-js-sdk API

### ISSUE THREAD

**Issue by Slapbox** [invertase/react-native-firebase#7483]
## Issue

**Unimplemented JS SDK features include:**

- storage.getBlob
- storage.getBytesSync
- database.forceLongPolling
- database.forceWebSockets

**Also possibly:**
- auth.unenroll
- auth.addTokenListener

I located these by searching for the string `not implemented` in the portions of the project we use. I've also made an effort to document the others but this may not be comprehensive, or may contain false positives.

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `e.g. 5.4.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `Y/N` & `VERSION`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by Slapbox** [comment]
Originally mentioned in #7433.

**Comment by Slapbox** [comment]
I also noticed `uploadBytes` is not implemented also. Is there anything else we can use in the meantime, or any plan to implement this in the immediate future? It would be a shame to have to revert adding `react-native-firebase` to our app. 

Unfortunately I lack the knowledge and time to help implement it.

**Comment by mikehardy** [comment]
@Slapbox we handle the non-modular `put` API which is very nearly `uploadBytes` by signature, it handles the same type of input which tells me that under the covers we are already doing all the necessary things from javascript to native and back

So my first take on it is that `uploadBytes` is probably just exporting the correct symbol and API shape, and should be nearly trivial. Those are famous last words but it really does look that way to me

Need to get react-native 0.73 unexpected incompatibility #7500  fixed before I can take a look at uploadBytes but I'll check it out with priority tomorrow

**Comment by Slapbox** [comment]
> Those are famous last words but it really does look that way to me

Too true!

I really appreciate your attention on this a lot. Thank you for the great project and the great assistance!

----

**Workaround:**

I just came back to update that it seems that [`uploadBytesResumable`](https://firebase.google.com/docs/reference/js/storage.md#uploadbytesresumable) is implemented and seems so far to be a viable workaround for us, but they return different types. We're just lucky to have an implementation that can readily support both without rewriting code. Always those famous last words though.

**Comment by mikehardy** [comment]
The implementation of uploadBytes (if necessary? Do let me know if uploadBytesResumable doesn't work out...) is likely to just internally do the UploadTask await for `then` -> then dereference final snapshot -> map final metadata + ref into an UploadResult -> return that

Which is to say it may reduce some boilerplate by handling that for you but it's just bending around the same information and should be working exactly the same under the covers in the native layer

**Comment by mikehardy** [comment]
Coming over from #4166 - 

getMultiFactorResolver is definitely implemented and working, TOTP I don't think was a focus of the original implementer (and in fact I think it was actually added as a firebase feature after the implementation of MFA here but my memory is a bit vague on the timing, apologies if wrong)

It appears uploadBytes is actually not a blocker now, is that correct?

Auth-related things obviously would be though - no one is going to change their well-considered hard-won MFA auth flows without a serious (and unwanted) battle - I understand that

So the MFA Resolver should kick out a list of hints and TOTP should be included (I think?) despite not being implemented yet here

At that point I think an implementation of `TotpMultiFactorGenerator` is all that is needed, and I don't recall that being very difficult.

Auth `unenroll` and token listeners are already present and working (in fact, native has a super-listener beyond the JS SDK of a "user change listener" which is the superset of token changes and user changes, but token listening by itself is I think all implemented.

So if I'm following along on your migration feasibility audit correctly it sounds like the TotpMultiFactorGenerator is really the blocker at the moment? If so, an implementation of that is something I could schedule

**Comment by Slapbox** [comment]
Thank you as always for your great work and assistance @mikehardy!

> It appears uploadBytes is actually not a blocker now, is that correct?

I'm 99% sure it's not a blocker, but I've not done extensive testing of every part of our application. We're doing a major rewrite so nothing is headed to production yet. Right now I'm just calling it "Good enough to keep moving in development."

If I encounter any portion of our code that can't use `uploadBytesResumable` as a drop-in replacement, it should be possible to adapt it on our end, but it would be good to have implemented in RNF eventually.

----

Regarding `TotpMultiFactorGenerator` - it's not a blocker for us at present, but in the future it would be if not implemented. We're planning to add TOTP support in the next few months, so I was scouting ahead to make sure the path is clear. I wanted to highlight that one because, whether we were going to use it or not, it's a security method without any possible workaround.

I think you're right that it didn't exist when `getMultiFactorResolver` was added. It's a rather new method; I'm pretty sure it's less than a year old.

**Comment by Slapbox** [comment]
Not stale.

**Comment by mikehardy** [comment]
Not only not stale, there's progress :-)

**Comment by Nantris** [comment]
Not stale.

**Comment by Nantris** [comment]
Bump.

**Comment by mikehardy** [comment]
Yep - usually not a fan of bumps but this actually tracks a very worthwhile chunk of work and I want it to stay open as well
I am not sure our "pin" actually avoids the stale bot (separate issue...) but hopefully it does.

**Comment by aem** [comment]
Curious if there's an update specifically on the `TotpMultiFactorGenerator` compatibility, and if that work can be prioritized. Unfortunately I don't personally have native mobile experience so I'm hesitant to try and contribute it, but not having access to TOTP second factors is tough given the relative insecurity of SMS.

**Comment by aem** [comment]
Would it maybe be helpful to split that out into its own issue so its not tied to the Storage + Database module work? I'm happy to do a deeper write up on the multi-factor work if that would help move it along

**Comment by hoongtong** [comment]
Is there any other ongoing development on adding TOTP compatibility aside from #7718  (which seems pretty stale)?

**Comment by mikehardy** [comment]
I have not seen it and am not aware of TOTP or MFA in general at current no - biggest blocker for me when I was developing more actively in this specific module is that the auth emulator does not support it and that still does not seem to be on their roadmap

It is very problematic - as a library developer - to run against the "real" / cloud auth APIs because they rate limit, they require a payment plan, the messages themselves always have some fakery since SMS cannot work, etc. We moved to the auth emulator a long time ago because the flakiness made developing + maintaining very difficult against the real APIs.

So, converting existing test suite back to real APIs, then fixing all the flakes etc so we have a stable baseline to develop with was the first hurdle and that's where things are at the moment. I suppose an alternate path could be to patch the auth emulator so it supports TOTP, that may be the easiest way to go in fact. Then implementing it here is probably pretty straightforward

**Comment by mikehardy** [comment]
Alternate thought - just picking up the PR and developing it in the form of local patches in an app where you know you can test it, until you know you have it working, then updating the PR and saying "hard to test but this works..." may be sufficient ?

**Comment by mare95** [comment]
Hi @mikehardy, you mentioned that it's possible to `unenroll` phone number as a second factor using react-native firebase library, right?

I can't find anything related to this in the docs.

I guess docs are not up to date or it has to be done via JS SDK? If so, then there is a bigger issue => `onIdTokenChanged` is not being triggered in the react-native app once i remove/add second factor via JS SDK which leads to incorrect state in the react-native app.

**Comment by mikehardy** [comment]
@mare95 unfortunately no it is not possible, there is (or was) a PR in #7566 here but it closed out due to being stale, and it mostly went stale for all the reasons mentioned above (that is: we use the auth emulator vs the auth cloud APIs in our e2e suite, but the auth emulator doesn't support MFA or TOTP so it is really really hard to support these APIs)

And you are correct, in that since the react-native-firebase library (which uses the native SDKs) will not be aware at all of any work you do in the firebase-js-sdk - so no listeners will fire. I think a workaround *could* be (needs testing!) to call the react-native-firebase user `reload` method after doing something in firebase-js-sdk auth, in order to force the native SDKs to consult the cloud User auth status at which point they should see anything that was changed by firebase-js-sdk and persisted to the cloud

**Comment by mikehardy** [comment]
As a user, if I enable 2FA via one channel then go to a session on another channel, I am not surprised at all to be logged out - in fact I think it would be very suspicious to leave a session open from before 2FA enable. If that's the scenario you describe I don't think there is a workaround and I think it is expected. Perhaps in the 2FA enable flow you could educate the user with a message similar to "Congratulations on your upgraded security! Any open sessions on other devices will be invalid after this change until you log in to them using your the stronger 2FA protection" or similar (sales! marketing! 😆 )

**Comment by Nantris** [comment]
@mikehardy I'm suspecting that using `uploadBytesResumable` instead of `uploadBytes` could be a problem when backgrounding the app. We're flushing everything as fast as we can when the app is backgrounded, and execution never makes it past `uploadBytesResumable`.

I can't prove that `uploadBytes` would act differently, but it seems reasonable to think it _might_ complete faster or avoid use of timers that `uploadBytesResumable` might use for something? Since I've exhausted everything else we can try I figured you might have some thoughts on this.

**Comment by hoongtong** [comment]
@mare95 were you able to implement the user reloading? We were also thinking of using both the js and native sdk, but have not found a way to pass the authenticated state from the js sdk to the native one yet.

**Comment by mare95** [comment]
@hoongtong nope, it's not possible to keep the session alive on a native side once you add/remove 2nd factor via JS SDK and vice versa. Instead we just reload firebase user in app focus and presented a the fancy "Session expired" modal if session is invalid. After that user gets logged out.

As Mike said above this is kind of expected behaviour since you are changing something very important privacy related so you don't want to keep other sessions active if that is the case. 

Btw we analysed analytic events related to 2FA state change vs logging out users on a different platforms because of that and it turned out that the number of people affected is very very small.

This is obviously not a solution but it might be a useful insight.

**Comment by Nantris** [comment]
Any update on (or interest in) implementing `uploadBytes`? I do believe this is causing our web sync logic to fail to complete before the app becomes backgrounded.

**Comment by artanisdesign** [comment]
getBlob would be nice, so we could move away from getDownloadURL..

**Comment by mikehardy** [comment]
Any APIs missing do have interest - the goal is always zero gaps, and we appreciate the patience. We'll accept PRs too of course 😅 

The last of the namespaced API / modular migration PRs are working through the process now, which is a vital internal task and has soaked up a lot of bandwidth. Hopefully we can close some of these feature gaps after

**Comment by jamespb97** [comment]
>
> * [ ]  **auth.TotpMultiFactorGenerator (Required for authenticator-app-based multifactor)**
> 
> **Also possibly the following:**
> 
> * [ ]  auth.unenroll[ ]  auth.addTokenListener
> 

These can now be checked 🥳

**Comment by Nantris** [comment]
@jamespb97 got 'em!

Thanks to everyone for the continued work on this library - in this case especially @jamespb97 and @mikehardy!

**Comment by mikehardy** [comment]
We're getting there. Did so much work on the library in the last month and it's in a good spot to get more APIs done, even with the AI package work taking some of our bandwidth 🚂
